dattri/algorithm/rps.py

```diff
@@ -0,0 +1,181 @@
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Callable
+
+    from torch import Tensor
+    from torch.utils.data import DataLoader
+
+import warnings
+
+import torch
+from torch.nn.functional import normalize
+
+from dattri.algorithm.utils import (
+    _check_shuffle,
+    get_rps_weight,
+    rps_finetune_theta,
+)
+from dattri.model_utils.hook import get_final_layer_io
+
+from .base import BaseAttributor
+
+
+class RPSAttributor(BaseAttributor):
+    """Representer point selection attributor."""
+
+    def __init__(
+        self,
+        loss_func: Callable,
+        model: torch.nn.Module,
+        final_linear_layer_name: str,
+        nomralize_preactivate: bool = False,
+        l2_strength: float = 0.003,
+        epoch: int = 3000,
+        device: str = "cpu",
+    ) -> None:
+        """Representer point selection attributor.
+
+        Args:
+            loss_func (Callable): The loss function to be attributed. The inputs are
+                list of pre-activation values (f_i in the paper) and list of labels.
+                Typical examples are BCELoss and CELoss.
+            model (torch.nn.Module): The model to attribute. RPS will extract
+                second-to-last layer results and the final fc layer's parameter. The
+                second one will be used for the initialization of the l2-finetuning.
+                That is, model output = fc(second-to-last feature).
+            final_linear_layer_name (str): The name of the final linear layer's name
+                in the model.
+            nomralize_preactivate (bool): If set to true, then the intermediate layer
+                output will be normalized. The value of the output inner-product will
+                not affected by the value of individual output magnitude.
+            l2_strength (float): The l2 regularizaton to fine-tune the last layer.
+            epoch (int): The number of epoch used to fine-tune the last layer.
+            device (str): The device to run the attributor. Default is cpu.
+        """
+        self.target_func = loss_func
+        self.model = model
+        self.final_linear_layer_name = final_linear_layer_name
+        self.nomralize_preactivate = nomralize_preactivate
+        self.l2_strength = l2_strength
+        self.epoch = epoch
+        self.device = device
+        self.full_train_dataloader = None
+
+    def cache(self, full_train_dataloader: DataLoader) -> None:
+        """Cache the full dataset for fine-tuning.
+
+        Args:
+            full_train_dataloader (DataLoader): The dataloader
+                with full training samples for the last linear layer
+                fine-tuning.
+        """
+        self.full_train_dataloader = full_train_dataloader
+
+        # get intermediate outputs and predictions for full train dataset
+        intermediate_x_train, y_pred_train = get_final_layer_io(
```

Comment:
Could this be too large for a large full_train_dataloader? Since we will have all the feature and output concanated together as 2 tensors.

Reply:
Yeah, you are right. If this is really an issue, we can probably do this in a batch manner?

Reply:
Now I think this should be minor? The peak mem of RPS is always much smaller than all the other methods.